akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipLatestWithApply.scala.template

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (C) 2014-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.stream.scaladsl
+
+import akka.stream._
+import akka.stream.impl.StreamLayout
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+
+trait ZipLatestWithApply {
+
+  [2..20#/**
+   * Create a new `ZipLatestWith` specialized for 1 inputs.
+   *
+   * @param zipper zipping-function from the input values to the output value
+   */
+  def apply[[#A1#], O](zipper: ([#A1#]) ⇒ O): ZipLatestWith1[[#A1#], O] =
+    new ZipLatestWith1(zipper)
+  #
+
+  ]
+
+}
+
+[2..20#/** `ZipLatestWith` specialized for 1 inputs */
+class ZipLatestWith1[[#A1#], O] (val zipper: ([#A1#]) ⇒ O) extends GraphStage[FanInShape1[[#A1#], O]] {
+  override def initialAttributes = Attributes.name("ZipLatestWith1")
+  override val shape: FanInShape1[[#A1#], O] = new FanInShape1[[#A1#], O]("ZipLatestWith1")
+  def out: Outlet[O] = shape.out
+  [#val in0: Inlet[A1] = shape.in0#
+  ]
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) { outer =>
+      // Without this field the completion signalling would take one extra pull
+      var willShutDown = false
+
+      [#val inlet0 = new ZipLatestInlet(in0)#
+      ]
+      var waitingForTuple = false
+      var waitingForNextTuple = false
+
+      override def preStart(): Unit = {
+        [#pull(in0)#
+        ]
+      }
+
+      setHandler(
+        out,
+        new OutHandler {
+          override def onPull(): Unit = {
+            if (hasAllValues) {
+                if (waitingForNextTuple) {
+                    waitingForTuple = true
+                } else {
+                    pushOutput()
+                }
+            }
+            else
+            {
+                waitingForTuple = true
+            }
+            pullAllIfNeeded()
+          }
+        }
+      )
+
+      [#setHandler(in0, inlet0)#
+      ]
+
+      private def hasAllValues = [#inlet0.hasValue#&&]
+
+      private def pushOutput(): Unit = {
+        push(out, zipper([#inlet0.value#,]))
+        if (willShutDown) completeStage()
+        waitingForNextTuple = true
+      }
+
+      private def pullAllIfNeeded(): Unit = {
+        [#if (!hasBeenPulled(in0)) {
+          pull(in0)
+        }#
+        ]
+      }
+
+      private class ZipLatestInlet[T](in: Inlet[T]) extends InHandler {
+         var value: T = _
+         var hasValue = false
+
+         override def onPush() = {
+              value = outer.grab(in)
+              hasValue = true
+              outer.waitingForNextTuple = false
+              if (outer.waitingForTuple && outer.hasAllValues) {
+                  outer.pushOutput()
+                  outer.waitingForTuple = false
+                  outer.pullAllIfNeeded()
+              }
+         }
+
+         override def onUpstreamFinish(): Unit = {
+           if (!outer.isAvailable(in)) outer.completeStage()
+           outer.willShutDown = true
```

Comment:
How does this work? Isn't there a risk that we will later `pull` from a finished inlet? Or is that guarded against via `hasBeenPulled`? Otherwise we might want to use `tryPull`.

Reply:
thanks for this comment, I have had a closer look at how termination is handled and realized that one case was not covered: the case when a first tuple is ready for emission and we have an upstream completion right after: in the zip operator, semantics are to still emit the first tuple. 

So I've changed the semantics slightly to match this behavior and added the corresponding coverage.

Regarding your question, I don't think this case can arise since whenever we have stale input values (which is also the case initially), we complete the stage immediately, and in the other case we will will complete upon the following pull.

But to be on the safe side and since the semantics are in essence closer to what it's supposed to do, I have changed the call to 'tryPull'

Reply:
thanks